Add Marko

tigt opened the pull request:
- Reuses the changes in https://github.com/matschik/component-party/pull/148
- Places Marko with the other resumable frameworks (if there’s other logic for how the frameworks are sorted, I’d love to know)

tigt committed:
WIP: add Marko

tigt committed:
WTF, vscode

tigt committed:
Consistent final newlines

LuLaValva commented on content/2-templating/6-conditional/marko/TrafficLight.marko:
```diff
@@ -0,0 +1,17 @@
+<p>
+  You must
+  <if=light === "red">STOP</if>
+  <if=light === "orange">SLOW DOWN</if>
```
These can be `else-if`

tigt replied:
Hm, that probably would show off Marko more. The React version was a bunch of `{light === "red" && "STOP"}`, so this was the most direct translation.

LuLaValva commented on content/2-templating/6-conditional/marko/TrafficLight.marko:
```diff
@@ -0,0 +1,17 @@
+<let/light = TRAFFIC_LIGHTS[lightIndex]/>
+
+<button onClick () {
+    lightIndex++;
```
This is fine because it's consistent with the others, but it could definitely be simplified to
```js
lightIndex = (lightIndex + 1) % TRAFFIC_LIGHTS.length;
```

LuLaValva commented on content/4-component-composition/1-props/marko/UserProfile.marko:
```diff
@@ -0,0 +1,18 @@
+  age: number | void;
+  favouriteColors: string[];
+  isAvailable: boolean;
+}
```
We don't have TS for any other examples, so I think we should probably remove lines 1-7

mlrawlings replied:
I think I agree.  We also don't support TS in the playground currently.

tigt replied:
All the other component frameworks that had PropTypes or TS support used it for this example — do you think we should follow up and re-add it once it works in the playground for comparison’s sake?

LuLaValva commented on content/4-component-composition/2-emit-to-parent/marko/AnswerButton.marko:
```diff
@@ -0,0 +1,7 @@
+export interface Input {
+  onYes: function;
+  onNo: function;
+}
+
```
Same as above, no TS so we can remove the `export interface Input`

mlrawlings replied:
```suggestion
```

LuLaValva commented on content/4-component-composition/2-emit-to-parent/marko/App.marko:
```diff
@@ -0,0 +1,7 @@
+<let/isHappy = true/>
+<p>Are you happy?</p>
+<AnswerButton
+  onYes () { isHappy = true }
```
My stylistic opinion is telling me that we should remove the space before the parentheses
```
onYes() { isHappy = true }
```

tigt replied:
Yeah, it’s a leftover habit from my last job’s eslint config — function definitions had the space, function calls didn’t

LuLaValva commented on content/4-component-composition/4-slot-fallback/marko/FunnyButton.marko:
```diff
@@ -0,0 +1,16 @@
+  transition: "transform 0.2s cubic-bezier(0.34, 1.65, 0.88, 0.925) 0s",
+  outline: "0",
+}>
+  <${input.renderBody}>
```
Adding this comment for further discussion, we still haven't decided whether this method is better than the if-else case. It may be harder to understand for Marko newbies

tigt replied:
It was new to _me_ for sure

tigt replied:
> Tigt — Today at 1:04 PM
> @DylanPiercey and @MichaelRawlings, what do you think? Should we advertise that fallback feature, or do you think you'd rather people not know about it?

<!-- -->

> DylanPiercey — Today at 1:05 PM
> People should know about it, we should probably advertise it.

LuLaValva commented on content/6-form-input/2-checkbox/marko/IsAvailable.marko:
```diff
@@ -0,0 +1,5 @@
+<input#is-available
+  type="checkbox"
+  checked:=isAvailable
```
```suggestion
  checked:=input.isAvailable
```

LuLaValva reviewed (commented):
Just some quick comments, but overall these look nice

mlrawlings commented on content/2-templating/6-conditional/marko/TrafficLight.marko:
```diff
@@ -0,0 +1,17 @@
+<let/light = TRAFFIC_LIGHTS[lightIndex]/>
+
+<button onClick () {
+    lightIndex++;
+    if (lightIndex >= TRAFFIC_LIGHTS.length) {
+      lightIndex = 0;
+    }
```
If we're going to do a longer version, we should do the exact same thing as the others
```suggestion
    if (lightIndex + 1 > TRAFFIC_LIGHTS.length - 1) {
      lightIndex = 0;
    } else {
      lightIndex++;
    }
```

The current version [doesn't work](https://markojs.com/playground/#NobwRAdghgtgpmAXGAlhAJnAHgOhlAJwGsB7MAGjAAcoAXACyTAHoBjEmKkiOCWgZ2ZpMufMTKV2fXrSb9adFKwAEU+coAqAJQCCAMT0BJAMIB9ADKGA4gAkNAZWUBeZcAA6YAnHQfyyjyQEUBAA5nC+-mAhXrweALoA3G4QADwANnC0zGkoIfS0hhjYzsoADMwAfMkpalk5ebQl2vpGZpa2DsD1+YUicZXJ1QBGAK60tNzK3MY5rETKABQAlMogycobyt0FRVgA1HtJEJvKKABmi9u9xRUuzQYmFtZ29jgZoQwra8cnW7k9uxKpSOJwAvutlKCKgA5bCNbYpZijcbcKqpKgVcz-RoofiIZQAEhA21BiIx1XJPwAmiQRsoYCN5BCUucnNtnE4XB4vD4wBV7BoAPIABUR5zRGxZZzZ2I5XLAgWCYQ8-PMgoA6soACIa6Fis4S5RSmUNOWRaJwWJ8qyC-VoskVMCguJAA) because the `lightIndex` doesn't get immediately updated.

mlrawlings replied:
Otherwise what Luke suggested would work as well.  We just don't want to update a value and then read from it.
```suggestion
    lightIndex = (lightIndex + 1) % TRAFFIC_LIGHTS.length;
```

mlrawlings commented on content/2-templating/6-conditional/marko/TrafficLight.marko:
```diff
@@ -0,0 +1,17 @@
+static const TRAFFIC_LIGHTS = ["red", "orange", "green"];
+<let/lightIndex = 0/>
+<let/light = TRAFFIC_LIGHTS[lightIndex]/>
```
```suggestion
<const/light = TRAFFIC_LIGHTS[lightIndex]/>
```
derived values use `const`, otherwise the `light` does not keep in sync with the `lightIndex`

mlrawlings commented on content/3-lifecycle/2-on-unmount/marko/Time.marko:
```diff
@@ -0,0 +1,6 @@
+<let/time = new Date()/>
+<lifecycle() {
+  onMount () { this.timer = setInterval(_ => time = new Date()), 1000) }
+  onDestroy () { clearInterval(this.timer) }
+}/>
```
```suggestion
<lifecycle
  onMount () { this.timer = setInterval(_ => time = new Date(), 1000) }
  onDestroy () { clearInterval(this.timer) }
/>
```

tigt replied:
Whoops, good catch — artifact from starting with `<effect() { … }/>` and then changing it to `<lifecycle>`

mlrawlings commented on content/4-component-composition/1-props/marko/UserProfile.marko:
```diff
@@ -0,0 +1,18 @@
+export interface Input {
+  name: string;
+  age: number | void;
+  favouriteColors: string[];
+  isAvailable: boolean;
+}
+
```
```suggestion
```

mlrawlings commented on content/6-form-input/3-radio/marko/PickPill.marko:
```diff
@@ -0,0 +1,22 @@
+function handleChange (event) {
+  picked = event.target.value;
+}
+
+<let/picked = "red"/>
```
```suggestion
<let/picked = "red"/>
<const/handleChange(event) {
  picked = event.target.value;
}/>
```

`function` at this point in the template would be parsed as a custom tag, and it can't be `static function` because we're assigning to the instance variable.

tigt replied:
I was _wondering_ why the syntax highlighter wasn’t upset

tigt committed:
Apply suggestions from code review

Everything except for the TS declarations since I think Dylan ought to weigh in

Co-authored-by: Michael Rawlings <mirawlings@ebay.com>
Co-authored-by: Luke LaValva <lukelavalva@gmail.com>

tigt committed:
Feedback: no space before method parens, no TS, else-if, ColorSelect works now

tigt committed:
Pre-bugbash updates

tigt committed:
This is probably why they want us to provide a linter

nnelgxorz committed:
Update Qwik logo (#147)

tigt committed:
Finishing touches

tigt committed:
Argle

tigt committed:
@rturnq feedback

dangkhoipro committed:
Update App.svelte (#141)

Fix horizontal scroll in small screen

ismailassa committed:
content: add context sample for vue3 (#151)

content: checking context for vue3

content: Changes for content Vue

JohnsonMao committed:
Add Component composition > Context for vue2 (#153)

matschik committed:
fix(react): fix context react example

liou666 committed:
feat(content): add render app > solid (#157)

* feat(content): add render app > solid

* fix: update Solid file sort

* doc: update readme

matschik committed:
fix: fix font rendering on navbar (#158)

JohnsonMao committed:
Add Component composition > Context for Qwik (#155)

mastastealth committed:
Ember "Context" Example (#160)

* Ember "Context" Example

* Forgot backing class

Commit:
Match examples to other frameworks

tigt committed:
Merge branch 'main' into add-marko
